File: 169-majority-element/majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int candidate = -1;
        int votes = 0;
        for (int i=0; i<nums.size(); i+=1){
            if (votes == 0){
                candidate = nums[i];
            }
            if (nums[i] == candidate){
                votes+=1;
            } else if (nums[i] != candidate){
                votes -= 1;    
            }
        }
        int count = 0;
        for (int i=0; i<nums.size(); i+=1){
            if (nums[i] == candidate){
                count += 1;
            }
        }

        if (count > nums.size()/2){
            return candidate;
        }
        return -8;
    }
};

File: 169-majority-element/majority-element.go
func majorityElement(nums []int) int {
    cand := nums[0]
    vote := 1
    for i := 1; i < len(nums); i++ {
        if nums[i] == cand {
            vote++
        } else if nums[i] != cand {
            vote --
        }

        if vote < 0 {
            cand = nums[i]
            vote = 0
        }
    }
    vote = 0
    for _, num := range nums {
        if num == cand {
            vote++
        }
    }

    if vote > (len(nums)/2) {
        return cand
    }
    return -1
}